### PR TITLE
jbuf: allocate mutex and lock also jbuf_debug()

### PR DESCRIPTION
**Pull request by cspiel1**
For the RX thread https://github.com/baresip/baresip/discussions/2445 `jbuf_debug()` should also be thread safe.

### CODE REVIEW

File: src/jbuf/jbuf.c [hunk 1]
```diff
@@ -188,7 +219,7 @@ int jbuf_alloc(struct jbuf **jbp, uint32_t min, uint32_t max)
 	DEBUG_INFO("alloc: delay=%u-%u frames\n", min, max);
 
 	jb->pt = -1;
-	err = mtx_init(&jb->lock, mtx_plain) != thrd_success;
+	err = mutex_alloc(&jb->lock);
```
Comment (sreimers) [inline src/jbuf/jbuf.c:222]: You can use mutex in destructor if you delay destructor with:

```c
jb = mem_zalloc(sizeof(*jb), NULL);
...
err = mutex_alloc(&jb->lock);
...
mem_destructor(jb, jbuf_destructor);
```

File: src/jbuf/jbuf.c [hunk 2]
```diff
@@ -188,7 +219,7 @@ int jbuf_alloc(struct jbuf **jbp, uint32_t min, uint32_t max)
...
 	if (err) {
 		err = ENOMEM;
```
Comment (sreimers) [inline src/jbuf/jbuf.c:224]: with mutex_alloc `err` needs no override.

File: src/jbuf/jbuf.c [hunk 3]
```diff
@@ -207,10 +205,13 @@ int jbuf_alloc(struct jbuf **jbp, uint32_t min, uint32_t max)
...
 
 out:
-	if (err)
+	if (err) {
 		mem_deref(jb);
-	else
+	}
+	else {
 		*jbp = jb;
+		mem_destructor(jb, jbuf_destructor);
```
Comment (sreimers) [inline src/jbuf/jbuf.c:214]: Needs to be set directly after mutex_alloc, before frame alloc.

Comment (cspiel1) [reply src/jbuf/jbuf.c:214]: `jbuf_destructor()` crashes if `mutex_alloc()` was not successful.

Comment (sreimers) [reply src/jbuf/jbuf.c:214]: Right, but not after err check.

**Comment by cspiel1** [comment]
Now I have a better solution. `jbuf_flush()` should check if the lock is NULL.

File: src/jbuf/jbuf.c (hunk 3 above)
Comment (sreimers) [reply src/jbuf/jbuf.c:215]:
```c
jb = mem_zalloc(sizeof(*jb), NULL);
... -no other allocations -
err = mutex_alloc(&jb->lock);
if (err)
   goto out;
mem_destructor(jb, jbuf_destructor);
```

**Comment by sreimers** [comment]
> Now I have a better solution. `jbuf_flush()` should check if the lock is NULL.

Thats not very reliable. Better fix allocation, so you can rely on mutex. So we handle this on other places too.

File: src/jbuf/jbuf.c (hunk 3 above)
Comment (cspiel1) [reply src/jbuf/jbuf.c]: Okay, or the newest version with NULL pointer check? Which one you prefer?

**Comment by cspiel1** [comment]
How do you like the variant how `jbuf_debug()` was made thread safe, with short mutex lock phase? Any further ideas?

```c
	struct mbuf *mb = mbuf_alloc(512);
	struct re_printf pfh = {
		vprintf_handler,
		mb
	};
```

File: src/jbuf/jbuf.c [hunk 4]
```diff
@@ -601,35 +611,49 @@ int jbuf_stats(const struct jbuf *jb, struct jbuf_stat *jstat)
 int jbuf_debug(struct re_printf *pf, const struct jbuf *jb)
 {
 	int err = 0;
+	struct mbuf *mb = mbuf_alloc(512);
+	struct re_printf pfh = {
+		vprintf_handler,
+		mb
+	};
```
Comment (alfredh) [inline src/jbuf/jbuf.c:618]: I think it is better to have straightforward and readable code.

I dont think that you get much performance benefit with this approach.

In general, clean and maintainable code should be preferred over performance.

Comment (cspiel1) [reply src/jbuf/jbuf.c:618]: @sreimers once wrote that printing to stdout may block sometimes.

I don't want to print to stdout while the mutex is locked. But I agree also with you and there might be a simpler solution with an mbuf. A commit will follow.

**Pull request merged**